ams/server/src/main/java/com/netease/arctic/server/manager/AbstractPluginManager.java

```diff
@@ -0,0 +1,318 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.server.manager;
+
+import com.alibaba.fastjson.JSONArray;
+import com.alibaba.fastjson.JSONObject;
+import com.netease.arctic.ams.api.ActivePlugin;
+import com.netease.arctic.server.Environments;
+import com.netease.arctic.server.exception.AlreadyExistsException;
+import com.netease.arctic.server.exception.LoadingPluginException;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.Yaml;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+
+/**
+ * Parent class of all plugin managers, which provide the common method to help load, install, and
+ * visit plugins.
+ *
+ * @param <T> The plugin types.
+ */
+public abstract class AbstractPluginManager<T extends ActivePlugin> implements PluginManager<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractPluginManager.class);
+  private static final String PLUGIN_CONFIG_DIR_NAME = "plugins";
+
+  private final Map<String, T> installedPlugins = new ConcurrentHashMap<>();
+  private final Map<String, T> foundedPlugins = new ConcurrentHashMap<>();
+  private final Map<String, PluginConfiguration> pluginConfigs = Maps.newConcurrentMap();
+  private final String pluginCategory;
+  private final Class<T> pluginType;
+  private final Executor pluginExecutorPool;
+
+  @SuppressWarnings("unchecked")
+  public AbstractPluginManager(String pluginCategory) {
+    this.pluginCategory = pluginCategory;
+    Type superclass = this.getClass().getGenericSuperclass();
+    Preconditions.checkArgument(
+        superclass instanceof ParameterizedType, "%s isn't parameterized", superclass);
+    pluginType = (Class<T>) ((ParameterizedType) superclass).getActualTypeArguments()[0];
+
+    // single thread pool, and min thread size is 1.
+    this.pluginExecutorPool =
```

Comment:
I suggest putting this thread pool in specific PluginManager such as EventManager

Reply:
> I suggest putting this thread pool in specific PluginManager such as EventManager

+1

Reply:
fixed